fairseq/search.py

```diff
@@ -75,11 +100,315 @@ def step(self, step: int, lprobs, scores: Optional[Tensor]):
+                k=each_k,
+                dim=1,
+            )
+            top_scores = top_scores.view(batch_size, -1)
+            top_indices = top_indices.view(batch_size, -1)
+            scores_buf = torch.cat((scores_buf, top_scores), dim=1)
+            indices_buf = torch.cat((indices_buf, top_indices), dim=1)
+            new_beams = torch.arange(0, beam_size, device=device).repeat(batch_size, 1)
+            beams_buf = torch.cat((beams_buf, new_beams), dim=1)
+
+        # Now, process sentences in the batch one by one.
+        new_scores_buf = torch.zeros((batch_size, 2 * beam_size), device=device)
+        new_indices_buf = torch.zeros((batch_size, 2 * beam_size), device=device).long()
+        new_beams_buf = torch.zeros((batch_size, 2 * beam_size), device=device).long()
+        for sentno, states in enumerate(constraint_states):
+            scores, indices, beams, new_states = self.step_sentence(step,
+                                                                    sentno,
+                                                                    lprobs[sentno],
+                                                                    constraint_states[sentno],
+                                                                    beams_buf[sentno].clone(),
+                                                                    indices_buf[sentno].clone(),
+                                                                    scores_buf[sentno].clone())
+            new_scores_buf[sentno] = scores
+            new_indices_buf[sentno] = indices
+            new_beams_buf[sentno] = beams
+            self.constraint_states[sentno] = new_states
+
+        return new_scores_buf, new_indices_buf, new_beams_buf
+
+    @torch.jit.export
+    def step_sentence(self,
+                      step: int,
+                      sentno: int,
+                      lprobs,
+                      constraint_states,
+                      beams_buf,
+                      indices_buf,
+                      scores_buf):
+        """Does per-sentence processing. Adds all constraints for each
+        hypothesis to the list of candidates; then removes duplicates,
+        sorts, and dynamically stripes across the banks.
+
+        :param step: The decoder timestep.
+        :param sentno: The sentence number in the batch.
+        :param lprobs: The target-vocab distributions for each item in the beam. Shape: (beam size, target vocab).
+        :param constraint_states: What constrained words have been generated by each item in the beam.
+        :param beams_buf: Beam numbers for each candidate. Shape: 1d list of cands.
+        :param indices_buf: Target vocab IDs for each candidate. Shape: 1d list of cands.
+        :param scores_buf: The score for each candidate. Shape: 1d list of scores.
+        """
+
+        device = lprobs.device
+
+        # STEP 2: Add all constraints for each beam item
+        for beamno, state in enumerate(constraint_states):
```

Comment:
I really enjoy pronouncing the variable name `beamno` out loud.

Reply:
Two syllables, or three?

Reply:
Two ("beam" + "no") :)